server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -713,6 +720,32 @@ private SearchPhaseResult executeQueryPhase(ShardSearchRequest request, SearchSh
         }
     }
 
+    public void executeRankFeaturePhase(RankFeatureShardRequest request, SearchShardTask task, ActionListener<RankFeatureResult> listener) {
```

Comment:
So, something that concerns me is that in this `SearchService` we have logic that will short cut things that are single shard-ed. Meaning, we will query & fetch in the same call to the shard. 

Do we handle this optimization anywhere for the rank phase?

Ah, I see 

```
            if (request.numberOfShards() == 1 && (request.source() == null || request.source().rankBuilder() == null)) {
                // we already have query results, but we can run fetch at the same time
                context.addFetchResult();
                return executeFetchPhase(readerContext, context, afterQueryTime);
            } 
```

We verify that the source doesn't have any rank builders. It would be good to add this optimization in the future I think right?

Do we have tests covering the single shard scenario to make sure our rank phase still runs?

Reply:
I think I see such tests later in the RankPhase tests classes.

Reply:
> It would be good to add this optimization in the future I think right?

++  This would be really nice :) I can follow up with another PR to avoid complicating (even more) this one :)